exporters/otlp/otlptrace/otlptracegrpc/client.go

```diff
@@ -17,94 +17,233 @@ package otlptracegrpc // import "go.opentelemetry.io/otel/exporters/otlp/otlptra
+	// made a mistake.
+	if c.tsc == nil {
+		return errNotStarted
+	}
+
+	// Clear c.tsc to signal the client is stopped.
+	c.tsc = nil
+
+	connErr := c.conn.Close()
+	if ctxErr != nil {
+		return ctxErr
+	}
+	return connErr
 }
 
-// UploadTraces sends a batch of spans to the collector.
+var errShutdown = errors.New("exporter is shutdown")
+
+// UploadTraces sends a batch of spans.
+//
+// Retryable errors from the server will be handled according to any
+// RetryConfig the client was created with.
 func (c *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.ResourceSpans) error {
-	if !c.connection.Connected() {
-		return fmt.Errorf("traces exporter is disconnected from the server %s: %w", c.connection.SCfg.Endpoint, c.connection.LastConnectError())
+	// Hold a read lock to ensure a shut down initiated after this starts does
+	// not abandon the export. This read lock acquire has less priority than a
+	// write lock acquire (i.e. Stop), meaning if the client is shutting down
+	// this will come after the shut down.
+	c.tscMu.RLock()
+	defer c.tscMu.RUnlock()
+
+	if c.tsc == nil {
+		return errShutdown
 	}
 
-	ctx, cancel := c.connection.ContextWithStop(ctx)
+	ctx, cancel := c.exportContext(ctx)
 	defer cancel()
-	ctx, tCancel := context.WithTimeout(ctx, c.connection.SCfg.Timeout)
-	defer tCancel()
-
-	ctx = c.connection.ContextWithMetadata(ctx)
-	err := func() error {
-		c.lock.Lock()
-		defer c.lock.Unlock()
-		if c.tracesClient == nil {
-			return errNoClient
-		}
-		return c.connection.DoRequest(ctx, func(ctx context.Context) error {
-			_, err := c.tracesClient.Export(ctx, &coltracepb.ExportTraceServiceRequest{
-				ResourceSpans: protoSpans,
-			})
-			return err
+
+	return c.requestFunc(ctx, func(iCtx context.Context) error {
+		_, err := c.tsc.Export(iCtx, &coltracepb.ExportTraceServiceRequest{
+			ResourceSpans: protoSpans,
 		})
+		// nil is converted to OK.
+		if status.Code(err) == codes.OK {
+			// Success.
+			return nil
+		}
+		return err
+	})
+}
+
+// exportContext returns a copy of parent with an appropriate deadline and
+// cancellation function.
+//
+// It is the callers responsibility to cancel the returned context once its
+// use is complete, via the parent of directly with the returned CancelFunc,to
+// ensure all resources are correctly released.
+func (c *client) exportContext(parent context.Context) (context.Context, context.CancelFunc) {
+	var (
+		ctx    context.Context
+		cancel context.CancelFunc
+	)
+
+	if c.exportTimeout > 0 {
+		ctx, cancel = context.WithTimeout(parent, c.exportTimeout)
+	} else {
+		ctx, cancel = context.WithCancel(parent)
+	}
+
+	if c.metadata.Len() > 0 {
+		ctx = metadata.NewOutgoingContext(ctx, c.metadata)
+	}
+
+	// Unify the client stopCtx with the parent.
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-c.stopCtx.Done():
+			// Cancel the export as the shutdown has timed out.
+			cancel()
+		}
 	}()
```

Comment:
Agreed, I was thinking about this as well. However, spawning the goroutine is the only way to avoid a deadlock when checking the context and continuing with the export at the same time. And [this is currently how this operation is already performed](https://github.com/open-telemetry/opentelemetry-go/blob/034956199a0bef5a1f3e0b81350f51a8fcf8df22/exporters/otlp/otlptrace/internal/connection/connection.go#L276-L290).

One thing I was thinking about is having a single goroutine that dynamically switches over registered contexts. That would limit the number of goroutines spawned. This would require some complex `reflect` usage as that is the only way to dynamically create a switch statement though. I decided against including that change in this PR.